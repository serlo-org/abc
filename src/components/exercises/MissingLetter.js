import { addIndex, map } from 'ramda';
import React from 'react';
import { View, Text } from 'react-native';

import speakerImage from '../../assets/images/speaker.png';

import { WHITE, PRIMARY } from '../../styles/colors';
import { DEFAULT } from '../../styles/text';
import { loadSounds } from '../helpers/audio';
import { RoundImageWithButton, TextPicker } from '../Components';
const mapIndexed = addIndex(map);

const MissingLetter = ({ image, sounds, text, missing, options }) => {
  const play = () => {
    const playAll = ([sound, ...rest]) => {
      sound.playAsync();
      sound.setPlaybackFinishedCallback(() => {
        sound.setPositionAsync(0);
        if (rest.length > 0) {
          setTimeout(() => playAll(rest), 1000);
        }
      });
    };
    playAll(sounds);
  };

  const letters = mapIndexed((char, key) => {
    return key === missing
      ? <View key={key}>
          <TextPicker options={options} />
        </View>
      : <View key={key} style={{ padding: 5 }}>
          <Text style={DEFAULT}>
            {char}
          </Text>
        </View>;
  }, text);

  return (
    <View
      style={{
        flex: 1,
        backgroundColor: PRIMARY,
        alignItems: 'center',
        justifyContent: 'space-around'
      }}
    >
      <RoundImageWithButton
        image={image}
        imageSize={200}
        icon={speakerImage}
        buttonSize={40}
        onPress={play}
      />
      <View style={{ flexDirection: 'row' }}>
        {letters}
      </View>
    </View>
  );
};

export default loadSounds(MissingLetter);
