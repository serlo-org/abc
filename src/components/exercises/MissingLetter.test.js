import React from 'react';

import MissingLetter from './MissingLetter';

import renderer from 'react-test-renderer';

it('renders without crashing', () => {
  const tree = renderer.create(
    <MissingLetter
      image={require('../../assets/images/apfel.jpg')}
      sounds={[
        require('../../assets/sounds/apfel_short.mp3'),
        require('../../assets/sounds/apfel_long.mp3')
      ]}
      text="Apfel"
      missing={3}
      options={['a', 'n', 'e']}
    />
  );
  expect(tree).toMatchSnapshot();
});
