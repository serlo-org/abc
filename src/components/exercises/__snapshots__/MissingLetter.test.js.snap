// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#00B4D5",
      "flex": 1,
      "justifyContent": "space-around",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.05)",
            "borderRadius": 9999999,
            "margin": 20,
          },
          undefined,
        ]
      }
    >
      <Image
        resizeMode="cover"
        source={1}
        style={
          Object {
            "borderRadius": 100,
            "height": 200,
            "margin": 20,
            "width": 200,
          }
        }
      />
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      isTVSelectable={true}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "marginLeft": -70,
          "marginRight": 20,
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Object {
            "backgroundColor": "#73DBFF",
            "borderRadius": 9999,
            "elevation": 10,
            "padding": 5,
            "shadowColor": "rgba(0, 0, 0, 0.1)",
            "shadowOffset": Object {
              "height": 4,
              "width": 4,
            },
            "shadowOpacity": 1,
            "shadowRadius": 0,
          }
        }
      >
        <Image
          source={1}
          style={
            Object {
              "height": 40,
              "resizeMode": "contain",
              "width": 40,
            }
          }
        />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "padding": 5,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#fff",
            "fontSize": 40,
            "fontWeight": "bold",
          }
        }
      >
        A
      </Text>
    </View>
    <View
      style={
        Object {
          "padding": 5,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#fff",
            "fontSize": 40,
            "fontWeight": "bold",
          }
        }
      >
        p
      </Text>
    </View>
    <View
      style={
        Object {
          "padding": 5,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#fff",
            "fontSize": 40,
            "fontWeight": "bold",
          }
        }
      >
        f
      </Text>
    </View>
    <View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "column",
            "justifyContent": "flex-end",
          }
        }
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          hitSlop={undefined}
          isTVSelectable={true}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <View
            style={
              Object {
                "backgroundColor": "#73DBFF",
                "borderRadius": 20,
                "elevation": 10,
                "padding": 5,
                "shadowColor": "rgba(0, 0, 0, 0.1)",
                "shadowOffset": Object {
                  "height": 4,
                  "width": 4,
                },
                "shadowOpacity": 1,
                "shadowRadius": 0,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#fff",
                  "fontSize": 40,
                  "fontWeight": "bold",
                  "textAlign": "center",
                }
              }
            >
              
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "padding": 5,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#fff",
            "fontSize": 40,
            "fontWeight": "bold",
          }
        }
      >
        l
      </Text>
    </View>
  </View>
</View>
`;
